// algs4/lib/Counter/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
/**
 *  The Counter class is a mutable data type to encapsulate a counter.
 *  <p>
 *  For additional documentation,
 *  see <a href="https://algs4.cs.princeton.edu/12oop">Section 1.2</a> of
 *  <i>Algorithms, 4th Edition</i> by Robert Sedgewick and Kevin Wayne.
 */
var Counter = /** @class */ (function () {
    function Counter(id) {
        this.name = id;
        this.count = 0;
    }
    /**
     * Increments the counter by 1.
     */
    Counter.prototype.increment = function () {
        this.count++;
    };
    /**
     * Returns the current value of this counter.
     *
     * @return the current value of this counter
     */
    Counter.prototype.tally = function () {
        return this.count;
    };
    /**
     * Returns a string representation of this counter.
     *
     * @return a string representation of this counter
     */
    Counter.prototype.toString = function () {
        return this.count + " " + this.name;
    };
    return Counter;
}());
exports.default = Counter;
