// algs4/lib/Bag/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
// helper linked list class
var Node = /** @class */ (function () {
    function Node() {
        this.item = null;
        this.next = null;
    }
    return Node;
}());
exports.Node = Node;
/**
 *  The Bag class represents a bag (or multiset) of
 *  generic items. It supports insertion and iterating over the
 *  items in arbitrary order.
 *  <p>
 *  This implementation uses a singly linked list with a static nested class Node.
 *  See {@link LinkedBag} for the version from the
 *  textbook that uses a non-static nested class.
 *  See {@link ResizingArrayBag} for a version that uses a resizing array.
 *  The <em>add</em>, <em>isEmpty</em>, and <em>size</em> operations
 *  take constant time. Iteration takes time proportional to the number of items.
 *  <p>
 *  For additional documentation, see <a href="https://algs4.cs.princeton.edu/13stacks">Section 1.3</a> of
 *  <i>Algorithms, 4th Edition</i> by Robert Sedgewick and Kevin Wayne.
 *
 *  @param <Item> the generic type of an item in this bag
 */
var Bag = /** @class */ (function () {
    /**
     * Initializes an empty bag.
     */
    function Bag() {
        this.first = null;
        this.n = 0;
        this.modCount = 0;
    }
    /**
     * Returns true if this bag is empty.
     *
     * @return `true` if this bag is empty;
     *         `false` otherwise
     */
    Bag.prototype.isEmpty = function () {
        return this.first == null;
    };
    /**
     * Returns the number of items in this bag.
     *
     * @return the number of items in this bag
     */
    Bag.prototype.size = function () {
        return this.n;
    };
    /**
     * Adds the item to this bag.
     *
     * @param  item the item to add to this bag
     */
    Bag.prototype.add = function (item) {
        this.modCount += 1;
        var oldfirst = this.first;
        this.first = new Node();
        this.first.item = item;
        this.first.next = oldfirst;
        this.n++;
    };
    /**
     * Returns an iterator that iterates over the items in this bag in arbitrary order.
     *
     * @return an iterator that iterates over the items in this bag in arbitrary order
     */
    Bag.prototype[Symbol.iterator] = function () {
        var _a = this, first = _a.first, modCount = _a.modCount;
        var current = first;
        var expectModCount = modCount;
        var self = this;
        return {
            next: function () {
                if (expectModCount !== self.modCount) {
                    throw new Error('ConcurrentModificationException');
                }
                if (current !== null) {
                    var result = {
                        done: false,
                        value: current.item,
                    };
                    current = current.next;
                    return result;
                }
                return {
                    done: true,
                    value: null,
                };
            },
        };
    };
    /**
     * Returns a string representation of this queue.
     *
     * @return the sequence of items in FIFO order, separated by spaces
     */
    Bag.prototype.toString = function () {
        var e_1, _a;
        var str = '';
        try {
            for (var _b = tslib_1.__values(this), _c = _b.next(); !_c.done; _c = _b.next()) {
                var node = _c.value;
                str += node + ", ";
            }
        }
        catch (e_1_1) { e_1 = { error: e_1_1 }; }
        finally {
            try {
                if (_c && !_c.done && (_a = _b.return)) _a.call(_b);
            }
            finally { if (e_1) throw e_1.error; }
        }
        return "Bag [" + str + "]";
    };
    return Bag;
}());
exports.default = Bag;
