// algs4/lib/Queue/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var Node = /** @class */ (function () {
    function Node() {
        this.item = null;
        this.next = null;
    }
    return Node;
}());
exports.Node = Node;
/**
 *  The `Queue` class represents a first-in-first-out (FIFO)
 *  queue of generic items.
 *  It supports the usual <em>enqueue</em> and <em>dequeue</em>
 *  operations, along with methods for peeking at the first item,
 *  testing if the queue is empty, and iterating through
 *  the items in FIFO order.
 *  <p>
 *  This implementation uses a singly linked list with a static nested class for
 *  linked-list nodes. See {@link LinkedQueue} for the version from the
 *  textbook that uses a non-static nested class.
 *  See {@link ResizingArrayQueue} for a version that uses a resizing array.
 *  The <em>enqueue</em>, <em>dequeue</em>, <em>peek</em>, <em>size</em>, and <em>is-empty</em>
 *  operations all take constant time in the worst case.
 *  <p>
 *  For additional documentation, see <a href="https://algs4.cs.princeton.edu/13stacks">Section 1.3</a> of
 *  <i>Algorithms, 4th Edition</i> by Robert Sedgewick and Kevin Wayne.
 *
 *  @param <Item> the generic type of an item in this queue
 */
var Queue = /** @class */ (function () {
    function Queue() {
        this.first = null;
        this.last = null;
        this.n = 0;
        this.modCount = 0;
    }
    /**
     * Returns true if this queue is empty.
     *
     * @return  `true` if this queue is empty; `false` otherwise
     */
    Queue.prototype.isEmpty = function () {
        return this.first == null;
    };
    /**
     * Returns the number of items in this queue.
     *
     * @return the number of items in this queue
     */
    Queue.prototype.size = function () {
        return this.n;
    };
    /**
     * Returns the item least recently added to this queue.
     *
     * @return the item least recently added to this queue
     * @throws Error if this queue is empty
     */
    Queue.prototype.peek = function () {
        if (this.isEmpty())
            throw new Error('Queue underflow');
        return this.first.item;
    };
    /**
     * Adds the item to this queue.
     *
     * @param tem the item to add
     */
    Queue.prototype.enqueue = function (item) {
        this.modCount += 1;
        var oldlast = this.last;
        this.last = new Node();
        this.last.item = item;
        this.last.next = null;
        if (this.isEmpty()) {
            this.first = this.last;
        }
        else {
            oldlast.next = this.last;
        }
        this.n++;
    };
    /**
     * Removes and returns the item on this queue that was least recently added.
     *
     * @return the item on this queue that was least recently added
     * @throws Error if this queue is empty
     */
    Queue.prototype.dequeue = function () {
        this.modCount += 1;
        if (this.isEmpty()) {
            throw new Error('Queue underflow');
        }
        var item = this.first.item;
        this.first = this.first.next;
        this.n--;
        if (this.isEmpty())
            this.last = null;
        return item;
    };
    /**
     * Returns a string representation of this queue.
     *
     * @return the sequence of items in FIFO order, separated by spaces
     */
    Queue.prototype.toString = function () {
        var e_1, _a;
        var str = '';
        try {
            for (var _b = tslib_1.__values(this), _c = _b.next(); !_c.done; _c = _b.next()) {
                var node = _c.value;
                str += node + ", ";
            }
        }
        catch (e_1_1) { e_1 = { error: e_1_1 }; }
        finally {
            try {
                if (_c && !_c.done && (_a = _b.return)) _a.call(_b);
            }
            finally { if (e_1) throw e_1.error; }
        }
        return "Queue [" + str + "]";
    };
    /**
     * Returns an iterator that iterates over the items in this queue in FIFO order.
     *
     * @return an iterator that iterates over the items in this queue in FIFO order
     */
    Queue.prototype[Symbol.iterator] = function () {
        var _a = this, first = _a.first, modCount = _a.modCount;
        var current = first;
        var expectModCount = modCount;
        var self = this;
        return {
            next: function () {
                if (expectModCount !== self.modCount) {
                    throw new Error('ConcurrentModificationException');
                }
                if (current !== null) {
                    var result = {
                        done: false,
                        value: current.item,
                    };
                    current = current.next;
                    return result;
                }
                return {
                    done: true,
                    value: null,
                };
            },
        };
    };
    return Queue;
}());
exports.default = Queue;
