// algs4/lib/VisualAccumulator/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var StdDraw_1 = tslib_1.__importDefault(require("../StdDraw"));
var VisualAccumulator = /** @class */ (function () {
    function VisualAccumulator() {
        this.total = 0;
        this.n = 0;
    }
    // constructor(trials: number, max: number) {
    // StdDraw.setXscale(0, trials);
    // StdDraw.setYscale(0, max);
    // StdDraw.setPenRadius(0.005);
    // }
    VisualAccumulator.prototype.addDataValue = function (value) {
        this.n++;
        this.total += value;
        StdDraw_1.default.setPenColor(StdDraw_1.default.DARK_GRAY);
        StdDraw_1.default.point(this.n, value);
        StdDraw_1.default.setPenColor(StdDraw_1.default.RED);
        StdDraw_1.default.point(this.n, this.mean());
    };
    VisualAccumulator.prototype.mean = function () {
        return this.total / this.n;
    };
    VisualAccumulator.prototype.toString = function () {
        return "n = " + this.n + ", mean = " + this.mean();
    };
    return VisualAccumulator;
}());
exports.default = VisualAccumulator;
